src/providers/hana/qgshananewconnection.cpp

```diff
@@ -0,0 +1,348 @@
+  connect( btnConnect, &QPushButton::clicked, this, &QgsHanaNewConnection::btnConnect_clicked );
+  connect( buttonBox, &QDialogButtonBox::helpRequested, this, &QgsHanaNewConnection::showHelp );
+
+  txtDriver->setText( QgsHanaDriver::instance()->getDriver() );
+
+  cbxCryptoProvider->addItem( tr( "openssl" ), QLatin1String( "openssl" ) );
+  cbxCryptoProvider->addItem( tr( "commoncrypto" ), QLatin1String( "commoncrypto" ) );
+  cbxCryptoProvider->addItem( tr( "sapcrypto" ), QLatin1String( "sapcrypto" ) );
+  cbxCryptoProvider->addItem( tr( "mscrypto" ), QLatin1String( "mscrypto" ) );
+
+  mAuthSettings->setDataprovider( QStringLiteral( "hana" ) );
+  mAuthSettings->showStoreCheckboxes( true );
+
+  if ( connName.isEmpty() )
+  {
+    rbtnSingleContainer->setChecked( true );
+    frmMultitenantSettings->setEnabled( false );
+  }
+  else
+  {
+    QgsHanaSettings settings( connName, true );
+    updateControlsFromSettings( settings );
+  }
+  txtName->setValidator( new QRegExpValidator( QRegExp( "[^\\/]*" ), txtName ) );
+
+  chkEnableSSL_clicked();
+}
+
+void QgsHanaNewConnection::accept()
+{
+  if ( isStringEmpty( txtName->text() ) )
+  {
+    QMessageBox::warning( this,
+                          tr( "Save Connection" ), tr( "WARNING: Connection name cannot be empty." ), QMessageBox::Ok );
+    return;
+  }
+
+  if ( isStringEmpty( txtDriver->text() ) )
+  {
+    QMessageBox::warning( this,
+                          tr( "Save Connection" ), tr( "WARNING: Driver field cannot be empty." ), QMessageBox::Ok );
+    return;
+  }
+
+  if ( isStringEmpty( txtHost->text() ) )
+  {
+    QMessageBox::warning( this,
+                          tr( "Save Connection" ), tr( "WARNING: Host field cannot be empty." ), QMessageBox::Ok );
+    return;
+  }
+
+  if ( rbtnMultipleContainers->isChecked() && rbtnTenantDatabase->isChecked() )
+  {
+    if ( isStringEmpty( txtTenantDatabaseName->text() ) )
+    {
+      QMessageBox::warning( this,
+                            tr( "Save Connection" ), tr( "WARNING: Tenant database name cannot be empty." ), QMessageBox::Ok );
+      return;
+    }
+  }
+
+  QString connName = txtName->text();
+  QgsHanaSettings::setSelectedConnection( connName );
+  bool hasAuthConfigID = !mAuthSettings->configId().isEmpty();
+
+  if ( !hasAuthConfigID && mAuthSettings->storePasswordIsChecked() &&
+       QMessageBox::question( this,
+                              tr( "Saving Passwords" ),
+                              tr( "WARNING: You have opted to save your password. It will be stored in unsecured "
+                                  "plain text in your project files and in your home directory (Unix-like OS) or user profile (Windows). "
+                                  "If you want to avoid this, press Cancel and either:\n\na) Don't save a password in the connection "
+                                  "settings  — it will be requested interactively when needed;\nb) Use the Configuration tab to add your "
```

Comment:
The hyphen is non-ASCII character. That means that we rely on the source code file being encoded as UTF-8 and be parsed as UTF-8. Probably you have just copied over the string from somewhere else. So it should be ok.

Reply:
Right, I have copied that from another provider.